### PR TITLE
Generalize categorical distribution related functions

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
Generalizes functions related to categorical distribution to accept general Eigen expressions.

## Tests
In some tests explicit templating is not required anymore.

## Side Effects
None.

## Release notes
Generalized functions related to categorical distribution to accept general Eigen expressions.

## Checklist

- [x] Math issue #1470

- [x] Copyright holder: Tadej Ciglarič
    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by andrjohns** [comment]
Hey Tadej, I'll review this and the cauchy PR tomorrow

File: stan/math/prim/err/check_simplex.hpp [hunk 1]
```diff
@@ -35,29 +35,28 @@ namespace math {
...
-template <typename T, int R, int C>
-void check_simplex(const char* function, const char* name,
-                   const Eigen::Matrix<T, R, C>& theta) {
-  using size_type = index_type_t<Eigen::Matrix<T, Eigen::Dynamic, 1>>;
+template <typename T, require_eigen_t<T>* = nullptr>
+void check_simplex(const char* function, const char* name, const T& theta) {
   using std::fabs;
   check_nonzero_size(function, name, theta);
-  if (!(fabs(1.0 - theta.sum()) <= CONSTRAINT_TOLERANCE)) {
+  ref_type_t<T> theta_ref = theta;
```
Comment (rok-cesnovar) [inline stan/math/prim/err/check_simplex.hpp:42]: We do this because theta is used in two places? To prevent multiple evaluations?

File: test/unit/math/prim/prob/categorical_log_test.cpp [hunk 2]
```diff
@@ -6,12 +6,12 @@ TEST(ProbDistributionsCategorical, log_matches_lpmf) {
...
-                  (stan::math::categorical_log<true, double>(1, theta)));
-  EXPECT_FLOAT_EQ((stan::math::categorical_lpmf<false, double>(1, theta)),
-                  (stan::math::categorical_log<false, double>(1, theta)));
-  EXPECT_FLOAT_EQ((stan::math::categorical_lpmf<double>(1, theta)),
-                  (stan::math::categorical_log<double>(1, theta)));
+  EXPECT_FLOAT_EQ((stan::math::categorical_lpmf<true>(1, theta)),
+                  (stan::math::categorical_log<true>(1, theta)));
+  EXPECT_FLOAT_EQ((stan::math::categorical_lpmf<false>(1, theta)),
+                  (stan::math::categorical_log<false>(1, theta)));
+  EXPECT_FLOAT_EQ((stan::math::categorical_lpmf<>(1, theta)),
```
Comment (rok-cesnovar) [inline test/unit/math/prim/prob/categorical_log_test.cpp:13]: Is this removal necessary or just removing to cleanup code? It might be nicer to have these without the empty `<>`

**Review by rok-cesnovar — COMMENTED** [review]
I want to help review these generalization PRs, so starting with some questions on this easy one. It looks good to me, just have 2 minor questions. One code style and another me not being sure of the generalization/expressions stuff.

File: stan/math/prim/err/check_simplex.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/prim/err/check_simplex.hpp:42]: yes

File: test/unit/math/prim/prob/categorical_log_test.cpp (hunk 2 above)
Comment (t4c1) [reply test/unit/math/prim/prob/categorical_log_test.cpp:13]: Now the template arguments have different meaning. So instantiating it with double would mean theta is double. Also now explicit template arguments are  never necessary. I can also delete the empty angle brackets.

Comment (rok-cesnovar) [reply test/unit/math/prim/prob/categorical_log_test.cpp:13]: Ok thanks. The cleanup is not really necessary, was more wondering if it was mandatory or something. Good to go.

Comment (t4c1) [reply test/unit/math/prim/prob/categorical_log_test.cpp]: It turns out some of these tests were duplicates. Fixed.

**Review by rok-cesnovar — APPROVED** [review]
